File: meeting_room2.cpp
// Time Complexity :O(nlogn)
// Space Complexity :O(k) : worst case(O(n))
// Did this code successfully run on Leetcode :-
// Any problem you faced while coding this :-

// Your code here along with comments explaining your approach
//1. Sort wrt to the starting time
//2. Implement min heap and compare end_time with minimum element
//3. Return the size of priority queue as it is the number of rooms required.
bool sortcol( const vector<int>& v1,const vector<int>& v2 ) {
    return v1[0] < v2[0];
}

class Solution {
public:
    
    int minMeetingRooms(vector<vector<int>>& intervals) {
        //edge
        if(intervals.size()==0) return 0;
        
        sort(intervals.begin(), intervals.end(),sortcol);
        //logic
        priority_queue <int, vector<int>, greater<>> min_heap;
        min_heap.push(0);
        int i=0;
        while(!min_heap.empty() && i<intervals.size()){
            int temp = min_heap.top();
            if(temp<=intervals[i][0]){
                min_heap.pop();
            }
            min_heap.push(intervals[i][1]);
            i++;
        }
        return (int)min_heap.size();
    }
        
    };

File: meeting room 2.java
// Time Complexity : o(nlogn) sort list and traverse it
// Space Complexity : o(k) k=no of rooms
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no




// Your code here along with comments explaining your approach

public class Solution {
    /**
     * @param intervals: an array of meeting time intervals
     * @return: the minimum number of conference rooms required
     */
    public int minMeetingRooms(List<Interval> intervals) {
        // Write your code here
        Collections.sort(intervals,(s1,s2)->s1.start-s2.start);
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        pq.add(intervals.get(0).end);
        for(int i=1;i<intervals.size();i++){
            int top=pq.peek();
            if(top<=intervals.get(i).start){
                pq.poll();
            }
            pq.add(intervals.get(i).end);
           
        }
        return pq.size();
    }
}